src/core/congestion_control.h

```diff
@@ -25,8 +44,10 @@ typedef struct QUIC_LOSS_EVENT {
 
     uint64_t LargestPacketNumberSent;
 
-    uint32_t NumRetransmittableBytes;
+    uint32_t NumAckedRetransmittableBytes;
```

Comment:
Why do we need to add a "number of ACKed bytes" field to QUIC_LOSS_EVENT? It looks like BbrCongestionControlOnDataLost is adding this to the current BytesInFlight as if it expects BytesInFlight to have already been decremented by the QUIC_ACK_EVENT, but the loss detection module already passes the loss event before the ack event to the CC- see this code comment in QuicLossDetectionProcessAckBlocks:

    if (NewLargestAck) {
        //
        // Handle packet loss (and any possible congestion events) before
        // data acknowledgement so that we have an accurate bytes in flight
        // calculation for congestion events.
        //
        QuicLossDetectionDetectAndHandleLostPackets(LossDetection, AckedRetransmittableBytes, (uint32_t)TimeNow);
    }

Reply:
`AckedRetransmittableBytes` is used to adjust the recovery window when BBR enters the fast recovery mode.

See `BBRModulateCwndForRecovery` function in [Modulating cwnd in Loss Recovery](https://tools.ietf.org/id/draft-cardwell-iccrg-bbr-congestion-control-00.html#modulating-cwnd-in-loss-recovery) part in the BBR doc.

Reply:
I still don't understand. What variable in the spec you linked to does QUIC_CONGESTION_CONTROL_BBR::RecoveryWindow correspond to?

Reply:
It's the `cwnd` in function `BBRModulateCwndForRecovery`.

The `RecoveryWindow` is the window size in "Fast Recovery" mode, and after we exit that mode, we'll continue use `Bbr->CongestionWindow`, which is what `BBRRestoreCwnd()` do in the spec.

Reply:
No, QUIC_CONGESTION_CONTROL_BBR::CongestionWindow is what corresponds to the "cwnd" variable in the spec. Not  QUIC_CONGESTION_CONTROL_BBR::RecoveryWindow.

I still think you should not be adding NumAckedRetransmittableBytes to BytesInFlight when you set RecoveryWindow. The logic in the spec's BBRModulateCwndForRecovery function assumes that packets_in_flight has already been decremented by packets_delivered during ACK processing (notice that BBRSetCwnd is all the way at the end of the set of operations in section 3.5.2, "per-ack steps"), but that isn't the case in the QUIC code.

Reply:
> QUIC_CONGESTION_CONTROL_BBR::CongestionWindow is what corresponds to the "cwnd" variable in the spec. Not QUIC_CONGESTION_CONTROL_BBR::RecoveryWindow.

Let say the "good cwnd" is "the cwnd unmodulated by loss recovery or ProbeRTT", which is `QUIC_CONGESTION_CONTROL_BBR::CongestionWindow` in our code.

As it described in the spec, the "good cwnd" will be saved to `BBR.prior_cwnd` before switch to recovery mode. And `BBR.cwnd` will be assigned the size of "RecoveryWindow".

After BBR exit the recovery mode, it calls `BBRRestoreCwnd()` to restore the previous "good cwnd" to `BBR.cwnd`.

Our implementation is essentially the same.

> I still think you should not be adding NumAckedRetransmittableBytes to BytesInFlight when you set RecoveryWindow...

Fixed.

Reply:
Now remove the NumAckedRetransmittableBytes field from QUIC_LOSS_EVENT, since it's no longer used.

Reply:
fixed